File: src/Arrays/LengthOfLongestSubstring3.java
package Arrays;

import java.util.HashMap;
import java.util.Map;

public class LengthOfLongestSubstring3 {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> windowMap = new HashMap<>();
        int left = 0, right = 0;
        int res = 0;
        while (right < s.length()) {
            char cr = s.charAt(right);
            right++;
            windowMap.put(cr, windowMap.getOrDefault(cr, 0) + 1);
            while (windowMap.getOrDefault(cr, 0) > 1) {
                char cl = s.charAt(left);
                left++;
                windowMap.put(cl, windowMap.get(cl) - 1);
            }
            res = Math.max(res, right - left);
        }
        return res;
    }

    public static void main(String[] args) {
        LengthOfLongestSubstring3 obj = new LengthOfLongestSubstring3();
        System.out.println(obj.lengthOfLongestSubstring("pwwkew"));
    }
}

File: go/slide_window/lengthOfLongestSubstring3.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	l := 0
	m := map[int]int{}
	res := 0
	for r, c := range s {
		m[int(c)]++
		for m[int(c)] > 1 {
			m[int(s[l])]--
			l++
		}
		res = max(res, r-l+1)
	}
	return res
}

func lengthOfLongestSubstring2(s string) int {
	l := 0
	m := [128]int{}
	res := 0
	for r, c := range s {
		m[c]++
		for m[c] > 1 {
			m[s[l]]--
			l++
		}
		res = max(res, r-l+1)
	}
	return res
}

func main() {

	// 2
	//fmt.Println(lengthOfLongestSubstring("aab"))
	// 3
	//fmt.Println(lengthOfLongestSubstring("abcabcbb"))
	// 3
	fmt.Println(lengthOfLongestSubstring2("pwwkew"))
}
